resources/Microsoft.VSCode.DSC/Microsoft.VSCode.DSC.psm1

```diff
@@ -0,0 +1,191 @@
+
+        $results = [VSCodeExtension[]]::new($extensionList.length)
+        
+        for ($i = 0; $i -lt $extensionList.length; $i++)
+        {
+            $extensionName, $extensionVersion = $extensionList[$i] -Split '@'
+            $results[$i] = [VSCodeExtension]@{
+                Name = $extensionName
+                Version = $extensionVersion
+            }     
+        }
+
+        return $results
+    }
+
+    [VSCodeExtension] Get()
+    {
+        $currentState = [VSCodeExtension]::InstalledExtensions[$this.Name]
+        if ($null -ne $currentState)
+        {
+            return $currentState
+        }
+        
+        return [VSCodeExtension]@{
+            Name = $this.Name
+            Version = $this.Version
+            Ensure = [VSCodeEnsure]::Absent
+        }
+    }
+
+    [bool] Test()
+    {
+        $currentState = $this.Get()
+        if ($currentState.Ensure -ne $this.Ensure)
+        {
+            return $false
+        }
+
+        if ($null -ne $this.Version -and $this.Version -ne $currentState.Version)
+        {
+            return $false
+        }
+
+        return $true
+    }
+
+    [void] Set()
+    {
+        if ($this.Test())
+        {
+            return
+        }
+
+        if ($this.Ensure -eq [VSCodeEnsure]::Present)
+        {
+            $this.Install($false)
+        }
+        else
+        {
+            $this.Uninstall($false)
+        }
+    }
+
+#region VSCodeExtension helper functions
+    static [void] GetInstalledExtensions()
+    {
+        [VSCodeExtension]::InstalledExtensions = @{}
+        foreach ($extension in [VSCodeExtension]::Export())
+        {
+            [VSCodeExtension]::InstalledExtensions[$extension.Name] = $extension
+        }          
+    }
+
+    [string] GetInstallArgument()
+    {
+        if ($null -eq $this.Version)
+        {
+            return $this.Name
+        }
+
+        return @(
+            $this.Name
+            $this.Version
+        ) -join '@'
+    }
+
+    [void] Install([bool] $preTest)
+    {
+        if ($preTest -and $this.Test())
+        {
+            return
+        }
+
+        Install-VSCodeExtension -Name $this.Name -Version $this.Version
+        [VSCodeExtension]::GetInstalledExtensions()
+    }
+
+    [void] Install()
+    {
+        $this.Install($true)
+    }
+
+    [void] Uninstall([bool] $preTest)
+    {
+        Uninstall-VSCodeExtension -Name $this.Name
+        [VSCodeExtension]::GetInstalledExtensions()
+    }
+
+    [void] Uninstall()
+    {
+        $this.Uninstall($true)
+    }
+#endregion VSCodeExtension helper functions
+}
+#endregion DSCResources
+
+function Install-VSCodeExtension
+{
+    [CmdletBinding()]
+    param (
+        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
+        [string]$Name,
+        [Parameter(ValueFromPipelineByPropertyName)]
+        [string]$Version
+    )
+        
+    Invoke-VSCode -Command "--install-extension $($this.GetInstallArgument())"
+}
```

Comment:
The `$this` variable in the function's scope will be `$null`, which I realize I must've missed when I was advising extraction - You could create a `Get-VSCodeExtensionInstallArgument` function to hold that logic in this function's `begin` block, and pass the `$PSBoundParameters` to it:

```suggestion
function Install-VSCodeExtension
{
    [CmdletBinding()]
    param (
        [Parameter(Mandatory, ValueFromPipelineByPropertyName)]
        [string]$Name,
        [Parameter(ValueFromPipelineByPropertyName)]
        [string]$Version
    )
    
    begin
    {
        function Get-InstallArgument
        {
            param([string]$Name, [string]$Version)

            if ([string]::IsNullOrEmpty($this.Version))
            {
                return $Name
            }

            return @(
                $Name
                $Version
            ) -join '@'
        }
    }
    
    process
    {
        $installArgument = Get-VSCodeExtensionInstallArgument @PSBoundParameters

        Invoke-VSCode -Command "--install-extension $installArgument"
    }
}
```